edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeAuthenticator.cs

```diff
@@ -81,13 +84,42 @@ async Task<(bool isAuthenticated, bool shouldFallback)> AuthenticateInternalAsyn
                     return (false, false);
                 }
 
-                bool syncServiceIdentity = this.syncServiceIdentityOnFailure && !reauthenticating;
-                (bool isAuthenticated, bool valueFound) = await this.AuthenticateWithServiceIdentity(tCredentials, tCredentials.Identity.Id, syncServiceIdentity);
-                if (!isAuthenticated && this.allowDeviceAuthForModule && tCredentials.Identity is IModuleIdentity moduleIdentity)
+                bool syncServiceIdentity = this.syncServiceIdentityOnFailure && !reauthenticating;
+                bool isAuthenticated = false;
+                bool valueFound = false;
+                string actorId = tCredentials.Identity.Id;
+
+                // Try to get the actorId from the credentials
+                (Option<string> actorDeviceId, Option<string> actorModuleId) = GetActorId(tCredentials);
+                if (actorDeviceId.HasValue)
+                {
+                    actorId = actorDeviceId.OrDefault() + actorModuleId.Map(id => "/" + id).GetOrElse(string.Empty);
+                }
+
+                if (this.nestedEdgeEnabled &&
+                    actorModuleId.Contains(Constants.EdgeHubModuleId) &&
+                    actorId != tCredentials.Identity.Id)
+                {
+                    // OnBehalfOf means an EdgeHub is trying to use its own auth to
+                    // connect as a child device or module. So if acting EdgeHub is
+                    // different than the target identity, then we know we're
+                    // processing an OnBehalfOf authentication.
+                    (isAuthenticated, valueFound) = await this.AuthenticateWithAuthChain(tCredentials, actorId, syncServiceIdentity);
+                }
+                else if (this.allowDeviceAuthForModule &&
+                    tCredentials.Identity is IModuleIdentity moduleIdentity &&
+                    !actorModuleId.HasValue)
                 {
-                    // Module can use the Device key to authenticate
+                    // We're trying to authenticate a module, but the token is for a device.
+                    // This is the scenario where a module can use its parent device's token
+                    // to authenticate.
                     Events.AuthenticatingWithDeviceIdentity(moduleIdentity);
-                    (isAuthenticated, valueFound) = await this.AuthenticateWithServiceIdentity(tCredentials, moduleIdentity.DeviceId, syncServiceIdentity);
+                    (isAuthenticated, valueFound) = await this.AuthenticateWithServiceIdentity(tCredentials, moduleIdentity.DeviceId, syncServiceIdentity, Option.None<string>());

```

Comment:
this needs to be failover still because when a module auth using the device's key, it's still using a module token.